File: backend/app/schemas/budget.py
"""Budget schemas"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from decimal import Decimal


class BudgetLineCreate(BaseModel):
    account_id: Optional[int] = None
    cost_center_id: Optional[int] = None
    jan: float = 0
    feb: float = 0
    mar: float = 0
    apr: float = 0
    may: float = 0
    jun: float = 0
    jul: float = 0
    aug: float = 0
    sep: float = 0
    oct: float = 0
    nov: float = 0
    dec: float = 0
    notes: Optional[str] = None


class BudgetCreate(BaseModel):
    fiscal_year: int = Field(..., ge=2000, le=2100)
    name: str = Field(..., min_length=1, max_length=100)
    branch_id: Optional[int] = None
    lines: List[BudgetLineCreate] = Field(default_factory=list)


class BudgetUpdate(BaseModel):
    name: Optional[str] = None
    status: Optional[str] = None


class BudgetLineResponse(BaseModel):
    id: int
    budget_id: int
    account_id: Optional[int]
    cost_center_id: Optional[int]
    jan: float
    feb: float
    mar: float
    apr: float
    may: float
    jun: float
    jul: float
    aug: float
    sep: float
    oct: float
    nov: float
    dec: float
    total: float
    notes: Optional[str]

    class Config:
        from_attributes = True


class BudgetResponse(BaseModel):
    id: int
    company_id: int
    branch_id: Optional[int]
    fiscal_year: int
    name: str
    status: str
    total_amount: Optional[float]
    approved_by: Optional[str]
    approved_at: Optional[datetime]
    created_by: Optional[str]
    created_at: datetime
    lines: Optional[List[BudgetLineResponse]] = None

    class Config:
        from_attributes = True


class BudgetVsActualLine(BaseModel):
    account_id: Optional[int]
    account_code: Optional[str]
    budgeted: float
    actual: float
    variance: float
    variance_percent: float


class BudgetVsActualResponse(BaseModel):
    budget_id: int
    budget_name: str
    fiscal_year: int
    lines: List[BudgetVsActualLine]
    summary: dict

File: frontend/src/lib/api/budget.ts
/**
 * Budget API Client
 * Handles all API calls for budget management
 */

import api from '@/lib/api';

export interface Budget {
  id: number;
  fiscal_year: number;
  name: string;
  status: 'draft' | 'approved' | 'archived';
  total_amount?: number;
  branch_id?: number;
  approved_by?: string;
  approved_at?: string;
  created_by?: string;
  created_at: string;
  lines?: BudgetLine[];
}

export interface BudgetLine {
  id: number;
  budget_id: number;
  account_id?: number;
  account_code?: string;
  account_name?: string;
  jan: number;
  feb: number;
  mar: number;
  apr: number;
  may: number;
  jun: number;
  jul: number;
  aug: number;
  sep: number;
  oct: number;
  nov: number;
  dec: number;
  total: number;
  notes?: string;
}

export interface BudgetVsActual {
  budget_id: number;
  budget_name: string;
  fiscal_year: number;
  lines: {
    account_id?: number;
    account_code?: string;
    budgeted: number;
    actual: number;
    variance: number;
    variance_percent: number;
  }[];
  summary: {
    total_budget: number;
    total_actual: number;
    utilization_percent: number;
  };
}

export interface CreateBudgetData {
  fiscal_year: number;
  name: string;
  branch_id?: number;
  lines?: {
    account_id?: number;
    cost_center_id?: number;
    [key: string]: any;
  }[];
}

export async function fetchBudgets(fiscalYear?: number, status?: string): Promise<Budget[]> {
  const params: any = {};
  if (fiscalYear) params.fiscal_year = fiscalYear;
  if (status) params.status = status;
  const response = await api.get('/api/budgets', { params });
  return response.data;
}

export async function fetchBudget(id: number): Promise<Budget> {
  const response = await api.get(`/api/budgets/${id}`);
  return response.data;
}

export async function createBudget(data: CreateBudgetData): Promise<Budget> {
  const response = await api.post('/api/budgets', data);
  return response.data;
}

export async function fetchBudgetVsActual(id: number): Promise<BudgetVsActual> {
  const response = await api.get(`/api/budgets/${id}/vs-actual`);
  return response.data;
}
